src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -71,6 +94,325 @@ private static Vector128<float> VectorSum(in Vector128<float> vector)
             }
         }
 
+        // Multiply matrix times vector into vector.
+        internal static unsafe void MatMulA(bool add, AlignedArray mat, AlignedArray src, AlignedArray dst, int crow, int ccol)
+        {
+            Contracts.Assert(Compat(mat));
+            Contracts.Assert(Compat(src));
+            Contracts.Assert(Compat(dst));
+
+            fixed (float* pSrcStart = &src.Items[0])
+            fixed (float* pDstStart = &dst.Items[0])
+            fixed (float* pMatStart = &mat.Items[0])
+            {
+                float* psrc = Ptr(src, pSrcStart);
+                float* pdst = Ptr(dst, pDstStart);
+                float* pmat = Ptr(mat, pMatStart);
+
+                float* pSrcEnd = psrc + ccol;
+                float* pDstEnd = pdst + crow;
+                float* pSrcCurrent = psrc;
```

Comment:
Why do we need another set of pointers here?  Can't we just use `psrc`?

Reply:
Actually, I think there is a bug here.  Shouldn't `pSrcCurrent` get reset to `psrc` after each inner loop?

From the C++ code:

```
        for (const float * ps = psrc; ps < psLim; ps += 4, pm += 4)
        {
            const float * pmTmp;
            __m128 x01 = _mm_load_ps(pmTmp = pm);
            __m128 x11 = _mm_load_ps(pmTmp += ccol);
            __m128 x21 = _mm_load_ps(pmTmp += ccol);
            __m128 x31 = _mm_load_ps(pmTmp += ccol);
            __m128 x02 = _mm_load_ps(ps);
```

Maybe we need a few more unit tests to cover different shapes of matrices.  We only have 4x4 and 4x8 tests.

---
In reply to: [208976581](https://github.com/dotnet/machinelearning/pull/668#discussion_r208976581) [](ancestors = 208976581)

Reply:
Reply to the first comment:
In this case, yes, we can use `psrc` directly and feel free to increment it in the for loops without having a `pSrcCurrent`, since we don't have to keep track of the start of the array (adjusted for the base of the `AlignedArray`) in this function.  The same goes for `dst` and `mat`.

In other functions that do not involve matrix multiplication, I have adopted the convention to only increment `pXCurrent`, while using `pX` to indicate a fixed pointer that indicates the start of the array `X`.  Should I skip using `pXCurrent` just for these 4-6 functions that involve matrix operations?